control-admin/src/main/java/fi/nls/oskari/control/admin/AbstractLayerAdminHandler.java

```diff
@@ -0,0 +1,140 @@
+import fi.nls.oskari.control.ActionException;
+import fi.nls.oskari.control.RestActionHandler;
+import fi.nls.oskari.domain.Role;
+import fi.nls.oskari.domain.User;
+import fi.nls.oskari.domain.map.OskariLayer;
+import fi.nls.oskari.log.LogFactory;
+import fi.nls.oskari.log.Logger;
+import fi.nls.oskari.service.OskariComponentManager;
+import fi.nls.oskari.service.ServiceException;
+import fi.nls.oskari.service.UserService;
+import fi.nls.oskari.util.JSONHelper;
+import fi.nls.oskari.util.PropertyUtil;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.oskari.permissions.PermissionService;
+import org.oskari.permissions.model.PermissionType;
+import org.oskari.permissions.model.Resource;
+import org.oskari.permissions.model.ResourceType;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public abstract class AbstractLayerAdminHandler extends RestActionHandler {
+
+    private Logger log = LogFactory.getLogger(AbstractLayerAdminHandler.class);
+    private PermissionService permissionsService;
+    private UserService userService;
+
+    private String[] capabilitiesRoles;
+    private Set<String> availablePermissionTypes;
+
+
+
+    public void init() {
+        capabilitiesRoles = PropertyUtil.getCommaSeparatedList("actionhandler.GetWSCapabilitiesHandler.roles");
+        permissionsService = OskariComponentManager.getComponentOfType(PermissionService.class);
+        availablePermissionTypes = getAvailablePermissions();
+
+    }
+
+    protected PermissionService getPermissionsService() {
+        return permissionsService;
+    }
+
+    protected UserService getUserService() throws ServiceException {
+        if (userService == null) {
+            userService = UserService.getInstance();
+        }
+        return userService;
+    }
+
+    protected boolean userHasEditPermission(User user, OskariLayer layer) {
+        return user.isAdmin() || permissionsService.findResource(ResourceType.maplayer, Integer.toString(layer.getId()))
+                .filter(r -> r.hasPermission(user, PermissionType.EDIT_LAYER)).isPresent();
+    }
+
+    protected boolean userHasAddPermission(User user) {
+        return user.isAdmin() || permissionsService.findResource(ResourceType.functionality, "generic-functionality")
+                .filter(r -> r.hasPermission(user, PermissionType.ADD_MAPLAYER)).isPresent();
+    }
+    protected boolean userHasCapabilitiesPermission(User user) {
+        return user.hasAnyRoleIn(capabilitiesRoles) || userHasAddPermission(user);
+    }
+    protected int getRoleId(String name) throws ServiceException {
+        final Role role = getUserService().getRoleByName(name);
+        return (int) role.getId();
+
+    }
+    protected JSONObject getPermissionsForLayer (User user, OskariLayer layer) throws ActionException, ServiceException {
+        if(!userHasEditPermission(user, layer)) throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
```

Comment:
This would be prettier wrapped in a `{ }` block

Reply:
Also I'm not sure if this is a correct place for permission checking. Seems like this function is called pretty deep into another function inside LayerAdminHandler. Also, the name of the function gives no clues about requiring the user to have an edit permission for the layer.

Reply:
In addition, while we will eventually write the permissions to the frontend (or store them somewhere else) as JSON, I really dislike mixing JSONObjects into functions dealing with logic and then passing them around the codebase. In my opinion it makes reading, maintaining and unit testing the code a lot more difficult. This function could easily be named `getPermissionsGroupByRole` returning a `Map<Role, List<String>>` and it would make it crystal clear what the function actually does. Then for JSON serialization we can later choose to use `Role#getName` if that's sufficient.